src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -781,24 +778,273 @@ namespace
 
         return StatusCode::Success;
     }
+
+    int initialize_context(const pal::string_t impl_dir, hostpolicy_contract &host_contract, const host_interface_t &host_interface, corehost_context_contract *context_contract)
+    {
+        pal::dll_t corehost;
+        int rc = hostpolicy_resolver::load(impl_dir, &corehost, host_contract);
```

Comment:
If I read the code correctly this means we end up loading the hostpolicy for each context separately.
Among other things this means we're calling corehost_load multiple times. Which is not really supported - I mean it works, because that's how tests are currently using it, but it doesn't do the right thing.
(It's also pretty weird, since the hostpolicy init structures are static and so potentially reinitializing them multiple times leads to weird races).

So following the logic here - maybe something like this would make sense:
* Remove the notion of "context" from hostpolicy public APIs - if we only ever have one there's no need to specify it.
* hostpolicy externally only holds one state - initialized and runtime loaded, or not (internally it would still keep the context to store the global variables on as that's a cleaner design I think).
* hostfxr must make sure to only ever call corehost_load onces (For the new APIs, the old APIs do end up calling it multiple times and I think we need to keep that - for the "command" thing to work).
* It would also mean that hostfxr would have to store the host_contract globally (or reuse it from the first context, basically same thing)

Just thinking out loud. In short the main complaint is that we should not be calling corehost_load more than once... and then the hostpolicy context thing. 
Note that the way it works now, there's no good place to do property unification for the secondary contexts - hostpolicy does not get the properties for the secondary context. Properties are passed through the corehost_load, which only every stores them the first time (subsequent calls are almost no-ops).

Reply:
Yes, we'd load the hostpolicy library once (then re-use it), but `corehost_load` would be called for each context (so multiple times). From the comment on that function about not reinitializing / returning success if already initialized, I had thought that calling `corehost_load` multiple times was okay/expected (but maybe that was weirdness introduced with com/ijw/winrt hosts?).

I had thought maybe `corehost_load` could at least return an error code indicating it was already initialized, but figured that would cause compat issues. The other thing I had considered was making the new APIs on hostfxr not call `corehost_load` at all and have `corehost_initialize_context` on hostpolicy handle conditionally doing the same thing internally (so similar outcome of only calling `corehost_load` once for the new APIs) - but that would be weirdly inconsistent with the existing APIs. Particularly for the component host scenarios, having the setting of that global init struct separate seemed a bit weird to me (e.g. in the existing code for getting a delegate for com et al, the first call to corehost_load wins with no indication for subsequent calls and then no information for that '[TODO] Validate the current CLR instance is acceptable for this request' about the current request)

After deciding to only track secondary contexts in hostfxr, I kept the 'context' concept on the hostpolicy public APIs with the thought that the limitation of only one 'context' was due to the runtime itself, not hostpolicy, so if that runtime limitation ever went away, hostpolicy APIs would still make sense. But maybe that is just really not going to happen?

I had made `corehost_initialize_context` take a `host_interface_t` with the intent of that being the way that it would get information about the secondary context (including properties) and check compatibility. (But not used yet, since I didn't deal with the compatibility checks yet..)

Reply:
Updated so that hostfxr should only call corehost_load if it hasn't already been done (for the new APIs).

Reply:
Removed 'context' from public hostpolicy APIs.

Reply:
I like the removed context - it is sort of still there, the `contract` could serve the same purpose eventually if we needed it (would be harder to implement as we would need effectively delegates, so maybe a bad idea).

As noted in my other comment, right now the most confusing thing about the first/secondary context init, is the `g_init` (and thus `corehost_load`) versus `g_context`. They seem to serve very similar purpose. But honestly good enough comments might make this mostly go away.